lms/djangoapps/courseware/views/views.py

```diff
@@ -1612,6 +1612,43 @@ def _track_successful_certificate_generation(user_id, course_id):
     })
 
 
+def enclosing_sequence_for_gating_checks(block):
+    """
+    Return the first ancestor of this block that is a SequenceDescriptor.
+
+    Returns None if there is no such ancestor. Returns None if you call it on a
+    SequenceDescriptor directly.
+
+    We explicitly test against the three known tag types that map to sequences
+    (even though two of them have been long since deprecated and are never
+    used). We _don't_ test against SequentialDescriptor directly because:
+
+    1. A direct comparison on the type fails because we magically mix it into a
+       SequenceDescriptorWithMixins object.
+    2. An isinstance check doesn't give us the right behavior because Courses
+       and Sections both subclass SequenceDescriptor. >_<
+
+    Also important to note that some content isn't contained in Sequences at
+    all. LabXchange uses learning pathways, but even content inside courses like
+    `static_tab`, `book`, and `about` live outside the sequence hierarchy.
+    """
```

Comment:
As usual, on behalf of all future code archaeologists, thanks for your comments which somehow make the wackiest code easy to comprehend.